# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/builtin/FinalizationRegistryObject.cpp
# Commit: 6f212eb26c1a
# Full Hash: 6f212eb26c1abd5a9a1de48d97a32548ae010a4f
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2020-06-23 21:38:40
# Regressor Bug: 1647325
# File Overlap Count: 1
# Description:
#   Bug 1647325 - Trace weak edges of FinalizationRegistrationsObjects when required r=sfink
#   
#   FinalizationRegistrationsObject only holds a vector of weak pointers to FinalizationRecordObjects, but it still needs a trace method so that those weak pointers can get updated by a moving GC.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D80630
# ==============================================================================

diff -r 63961738f91e -r 6f212eb26c1a js/src/builtin/FinalizationRegistryObject.cpp
--- a/js/src/builtin/FinalizationRegistryObject.cpp	Tue Jun 23 18:39:57 2020 +0000
+++ b/js/src/builtin/FinalizationRegistryObject.cpp	Tue Jun 23 15:22:58 2020 +0000
@@ -154,7 +154,7 @@
     nullptr,                                    // call
     nullptr,                                    // hasInstance
     nullptr,                                    // construct
-    nullptr,                                    // trace
+    FinalizationRegistrationsObject::trace,     // trace
 };
 
 /* static */
@@ -178,9 +178,20 @@
 }
 
 /* static */
+void FinalizationRegistrationsObject::trace(JSTracer* trc, JSObject* obj) {
+  if (!trc->traceWeakEdges()) {
+    return;
+  }
+
+  auto* self = &obj->as<FinalizationRegistrationsObject>();
+  TraceRange(trc, self->records()->length(), self->records()->begin(),
+             "FinalizationRegistrationsObject records");
+}
+
+/* static */
 void FinalizationRegistrationsObject::finalize(JSFreeOp* fop, JSObject* obj) {
-  auto rv = &obj->as<FinalizationRegistrationsObject>();
-  fop->delete_(obj, rv->records(), MemoryUse::FinalizationRecordVector);
+  auto* self = &obj->as<FinalizationRegistrationsObject>();
+  fop->delete_(obj, self->records(), MemoryUse::FinalizationRecordVector);
 }
 
 inline WeakFinalizationRecordVector*
@@ -348,7 +359,9 @@
     registrations->trace(trc);
   }
 
-  // The active record set is weakly held and is not traced.
+  // The active record set is weakly held and is not traced. For moving GC this
+  // is updated in sweep(), which is called for all FinalizationRegistryObjects
+  // in a zone.
 
   if (FinalizationRecordVector* records = registry->recordsToBeCleanedUp()) {
     records->trace(trc);